discord/voice_client.py

```diff
@@ -640,3 +687,158 @@ def send_audio_packet(self, data, *, encode=True):
+            return
+
+        self.decoder.decode(data)
+
+    def start_recording(self, sink, callback, *args):
+        """The bot will begin recording audio from the current voice channel it is in.
+        This function uses a thread so the current code line will not be stopped.
+
+        Must be in a voice channel to use.
+        Must not be already recording.
+
+        Parameters
+        ----------
+        sink: :class:`Sink`
+            A Sink which will "store" all the audio data
+        callback: :class:`asynchronous function`
+            A function which is called after the bot has stopped recording.
+        *args:
+            Args which will be passed to the callback function.
+
+        Raises
+        ------
+        ClientException
+            Not connected to a voice channel.
+        ClientException
+            Already recording.
+        ClientException
+            Must provide a Sink object.
+        """
+        if not self.is_connected():
+            raise ClientException('Not connected to voice channel.')
+        if self.recording:
+            raise ClientException("Already recording.")
+        if not isinstance(sink, Sink):
+            raise ClientException("Must provide a Sink object.")
+
+        sink.init()
+
+        t = threading.Thread(target=self.recv_audio, args=(sink, callback, *args,))
+        t.start()
+
+    def stop_recording(self):
+        """Stops the recording.
+
+        Must be already recording.
+
+        Raises
+        ------
+        ClientException
+            Not currently recording.
+        """
+        if not self.recording:
+            raise ClientException("Not currently recording audio.")
+        self.decoder.stop()
+        self.recording = False
+        self.paused = False
+
+    def pause_recording(self):
+        """Pauses or unpauses the recording.
+
+        Must be already recording.
+
+        Raises
+        ------
+        ClientException
+            Not currently recording.
+         """
+        if not self.recording:
+            raise ClientException("Not currently recording audio.")
+        self.paused = {True: False, False: True}[self.paused]
+
+    def continue_recv(self):
+        # If data is not received while not recording
+        # then it will be received when the next
+        # recording starts. This function is run at
+        # the end of recv_audio and will end its
+        # execution after another recording starts
+        while not self.recording:
+            try:
+                self.socket.recv(4096)
+            except OSError:
+                # Disconnected
+                return
+
+    def recv_audio(self, sink, callback, *args):
+        #  Gets data from _recv_audio and sorts
+        #  it by user, handles pcm files and
+        #  silence that should be added.
+        if not self.decoder:
+            self.decoder = opus.DecodeManager(self)
+        self.decoder.start()
```

Comment:
This line throws an exception if start_recording() is called more than once:

```
Exception in thread Thread-12:
Traceback (most recent call last):
  File "/usr/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.7/dist-packages/discord/voice_client.py", line 801, in recv_audio
    self.decoder.start()
  File "/usr/lib/python3.7/threading.py", line 843, in start
    raise RuntimeError("threads can only be started once")
RuntimeError: threads can only be started once

Exception in thread Thread-11:
Traceback (most recent call last):
  File "/usr/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/lib/python3.7/dist-packages/discord/sink.py", line 69, in wait_and_stop
    self.vc.stop_recording()
AttributeError: 'NoneType' object has no attribute 'stop_recording'
```

Reply:
Thanks, I've fixed it now and the updated code is there

Reply:
The latest commit fixes the exception, however it mixes some of the audio from a previous recording into subsequent recordings. For example, if I record a 3 second clip where I say "1 2 3" and then record a second clip where I say "a b c" I'll get a 4 second clip where I hear " 3 a b c".

Reply:
Ok I'll look into it tomorrow and see if I can fix it.

Reply:
Ok I think I fixed the issue. Discord is still sending audio over the socket while the program is not receiving, so it builds up and you end up getting audio from before you start listening. To fix it I had the socket drop all the packets built up after it starts.